pkg/decoder/decoder.go

```diff
@@ -84,25 +136,105 @@ func (d *Decoder) sendBuffuredMessage(offset int64) {
+		d.sendBufferedMessage(offset - int64(1+len(newLine))) // remove '\n' + new line length
+		d.msgBuffer.Write(newLine)
+		return
+	}
+
+	maxLen := maxMessageLen - d.msgBuffer.Len() - 2 
+	if len(newLine) >= maxLen {
+		d.msgBuffer.Write([]byte(`\n`))
+		d.msgBuffer.Write(newLine[:maxLen])
+		d.msgBuffer.Write(truncatedMsg)
+		d.sendBufferedMessage(offset - int64(1+len(newLine[maxLen:]))) // remove '\n' + the length of the piece of the line that can't fit
+		d.msgBuffer.Write(truncatedMsg)
+		d.msgBuffer.Write(newLine[maxLen:])
+		return
+	}
+
+	// append a new line to the message
+	d.msgBuffer.Write([]byte(`\n`))
+	d.msgBuffer.Write(newLine)
+}
+
+// recoverTooLongBufferedLine truncates the new line and sends its left part and the message if needed
+func (d *Decoder) recoverTooLongBufferedLine(offset int64) {
+	newLine := make([]byte, d.lineBuffer.Len())
+	copy(newLine, d.lineBuffer.Bytes())
+	defer d.lineBuffer.Reset()
+
+	// truncate and send new line
+	truncAndSendLine := func() {
+		newLine = append(newLine, truncatedMsg...)
+		d.msgBuffer.Write(newLine)
+		d.sendBufferedMessage(offset)
+		d.msgBuffer.Write(truncatedMsg)
+	}
+
+	if d.singleline || d.msgBuffer.Len() == 0 {
+		truncAndSendLine()
+		return
+	}
+
+	if d.re.Match(newLine) {
+		d.sendBufferedMessage(offset - int64(len(newLine)))
+		truncAndSendLine()
+		return
+	}
+
+	maxLen := maxMessageLen - d.msgBuffer.Len() - 1
+	d.msgBuffer.Write([]byte(`\n`))
+	d.msgBuffer.Write(newLine[:maxLen])
+	d.msgBuffer.Write(truncatedMsg)
+	d.sendBufferedMessage(offset - int64(len(newLine[maxLen:]))) // remove the length of the line that can't fit
+	d.msgBuffer.Write(truncatedMsg)
+	d.msgBuffer.Write(newLine[maxLen:])
+}
+
+// decodeIncomingData splits raw data based on `\n`, creates and processes new lines
+// returns true and the new offset if the end of inBuf is likely to be the end of a message
+func (d *Decoder) decodeIncomingData(inBuf []byte, offset int64) (ok bool, newOffset int64) {
 	var i, j = 0, 0
-	var maxj = maxMessageLen - d.msgBuffer.Len()
+	var maxj = maxMessageLen - d.lineBuffer.Len()
 	// Note: we will truncate messages of length MaxLen - truncatedLen
 	// instead of MaxLen. We'll live with it for now
 	for ; j < len(inBuf); j++ {
 		if inBuf[j] == '\n' {
-			d.msgBuffer.Write(inBuf[i:j])
-			d.sendBuffuredMessage(offset + int64(j+1))
+			d.lineBuffer.Write(inBuf[i:j])
+			d.processNewLine(offset + int64(j+1))
 			i = j + 1 // +1 as we skip the `\n`
-			maxj = maxMessageLen - d.msgBuffer.Len()
+			maxj = maxMessageLen - d.lineBuffer.Len()
 		} else if j == maxj {
-			d.msgBuffer.Write(inBuf[i:j])
-			d.msgBuffer.Write(truncatedMsg)
-			d.sendBuffuredMessage(offset + int64(j))
-			d.msgBuffer.Write(truncatedMsg)
+			d.lineBuffer.Write(inBuf[i:j])
+			d.recoverTooLongBufferedLine(offset + int64(j))
 			i = j
+			maxj = maxMessageLen - d.lineBuffer.Len()
 		}
 	}
-	d.msgBuffer.Write(inBuf[i:j])
+	d.lineBuffer.Write(inBuf[i:j])
+
+	if len(inBuf) > 0 && inBuf[j-1] == '\n' {
+		ok = true
```

Comment:
let's rename `ok` here too

Reply:
It's a tough one, I wanted to use the `ok, error` pattern of go but I agree it does not make that much sense. Would `isEndOfLine` sound better ?

Reply:
`endsWithLF`?